presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSampleFunction.java

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import com.facebook.presto.common.block.ArrayBlock;
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.RunLengthEncodedBlock;
+import com.facebook.presto.common.type.BigintType;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.operator.aggregation.NullablePosition;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.AggregationState;
+import com.facebook.presto.spi.function.BlockIndex;
+import com.facebook.presto.spi.function.BlockPosition;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.common.type.StandardTypes.BIGINT;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.max;
+
+@AggregationFunction(value = "reservoir_sample", isCalledOnNullInput = true)
+public class ReservoirSampleFunction
+{
+    public static final String NAME = "reservoir_sample";
+
+    private ReservoirSampleFunction()
+    {
+    }
+
+    @InputFunction
+    @TypeParameter("T")
+    public static void input(
+            @TypeParameter("T") Type type,
+            @AggregationState ReservoirSampleState state,
+            @BlockPosition @SqlType("array(T)") @NullablePosition Block initialState,
+            @BlockIndex int initialStatePos,
+            @SqlType(BIGINT) long initialProcessedCount,
+            @BlockPosition @SqlType("T") @NullablePosition Block value,
+            @BlockIndex int position,
+            @SqlType(BIGINT) long desiredSampleSize)
+    {
+        checkArgument(desiredSampleSize > 0, "desired sample size must be > 0");
+        if (initialProcessedCount <= 0) {
+            // initial state block must be null or empty to prevent confusing situation where the
+            // initial sample is not used
+            checkArgument(initialState.isNull(initialStatePos) || initialState.getBlock(initialStatePos).getPositionCount() == 0, "initial state array must be null or empty when initial processed count is <= 0");
+        }
+        if (state.get() == null) {
+            state.set(new ReservoirSample(type));
+        }
+        ReservoirSample sample = state.get();
+        sample.tryInitialize((int) desiredSampleSize);
```

Comment:
Since the current internal representation is an arraylist, it must be an `int`. I will update the function signature such that we accept the `INTEGER` type rather than `BIGINT`, however, we still need this cast here as the annotation framework doesn't support `int`-type arguments.

Reply:
updated